tower/src/load/completion.rs

```diff
@@ -0,0 +1,91 @@
+//! Application-specific request completion semantics.
+
+use futures_core::ready;
+use pin_project::pin_project;
+use std::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+/// Attaches `H`-typed completion tracker to `V` typed values.
+///
+/// Handles (of type `H`) are intended to be RAII guards that primarily implement `Drop` and update
+/// load metric state as they are dropped. This trait allows implementors to "forward" the handle
+/// to later parts of the request-handling pipeline, so that the handle is only dropped when the
+/// request has truly completed.
+///
+/// This utility allows load metrics to have a protocol-agnostic means to track streams past their
+/// initial response future. For example, if `V` represents an HTTP response type, an
+/// implementation could add `H`-typed handles to each response's extensions to detect when all the
+/// response's extensions have been dropped.
+///
+/// A base `impl<H, V> TrackCompletion<H, V> for CompleteOnResponse` is provided to drop the handle
+/// once the response future is resolved. This is appropriate when a response is discrete and
+/// cannot comprise multiple messages.
+///
+/// In many cases, the `Output` type is simply `V`. However, `TrackCompletion` may alter the type
+/// in order to instrument it appropriately. For example, an HTTP `TrackCompletion` may modify the
+/// body type: so a `TrackCompletion` that takes values of type `http::Response<A>` may output
+/// values of type `http::Response<B>`.
+pub trait TrackCompletion<H, V>: Clone {
+    /// The instrumented value type.
+    type Output;
+
+    /// Attaches a `H`-typed handle to a `V`-typed value.
+    fn track_completion(&self, handle: H, value: V) -> Self::Output;
+}
+
+/// A `TrackCompletion` implementation that considers the request completed when the response
+/// future is resolved.
+#[derive(Clone, Copy, Debug)]
+pub struct CompleteOnResponse;
```

Comment:
impl default and a `_p: ()` for backwards compat?

Reply:
The naming change already means it's not backwards compatible :p
Why is it you want a private field? To avoid people being able to match on it? Could we use `#[non_exhaustive]` instead?

Reply:
Yeah, just seems like a common pattern that we do in general 🤷

Reply:
`#[non_exhaustive]` _would_ bump msrv to 1.40, but I think that's generally the preferred way to do this?

Reply:
FWIW, `_p: ()` does more than _just_ preventing matching --- it also prevents user code from writing out a struct literal and requires the use of a constructor to get an instance of the type. 

In general, I think this can be a good practice for structs with no fields if there is the potential of them growing a field in the future, since adding the field would break any existing struct literals. If we can guarantee with some confidence that we won't need to ever add fields to this, it may not be necessary.

Reply:
Also (I'm not _sure_ about this OTTOMH) I _think_ that `#[non_exhaustive]` is only meaningful when applied to enum types. I think it specifically effects how _variants_ are matched. I don't know if it does anything here.

Reply:
`#[non_exhaustive]` does all of the things you listed above, and it works for anything "struct-like" (including unit structs) :)

I _do_ think that there may be a decent argument for us never adding fields to `CompleteOnResponse` though, in which case the annotation wouldn't be needed.

Reply:
> `#[non_exhaustive]` does all of the things you listed above, and it works for anything "struct-like" (including unit structs)

TIL! Never mind!